### ISSUE TITLE
[FEATURE] Generalizing `image_stack` for deep categorical inspection

### ISSUE THREAD

**Issue by jlstevens** [bokeh/bokeh#13354]
### Problem description

The new `image_stack` glyph allows for client-side color mixing, replacing the sort of operation that tools such as datashader would do on the server-side for visualizing categorical data (an exciting new feature by the way!).

One of the exciting features of being able to do this client-side is that you can use the hover tool to inspect image stacks. The work in https://github.com/bokeh/bokeh/pull/13193 address this partially with further (related!) discussion in https://github.com/bokeh/bokeh/issues/13200. However, even if both these suggestions were implemented, I think some of the potential power of image stack inspection would still not be realized. This issue outlines an ambitious vision for how I can imagine image stacks working.

Here is a mockup I've made for my idea hover inspection based on a real example (ship traffic around Vancouver):

<img width="1162" alt="image" src="https://github.com/bokeh/bokeh/assets/890576/a0874954-944e-4059-8d61-d9df5260af4f">

This would be an `image_stack` on top of a tile source where 'Cargo' and 'Passenger' are ship categories. The color mixing is based on the number of ships (of different categories) aggregated in a given pixel (the 'count' in datashader terminology). However, the data has a lot more structure than this! We also have the ships' name, tonnage, length and heading and what I think would be really useful is to see a hover tool with all this information for all the relevant categories at the cursor position (in practice this would be showing the output of a filter, e.g. the biggest ship by category, or some statistics aggregated over all the ships within each category).

To make this work, you would need to allow image stacks to be composed of an arbitrary number (`Z`) of `X x Y x C` ndarrays (where `X` and `Y` are the x- and y-resolutions and `C` is the number of categories). Of these `Z` ndarrays, only one would be used for color mixing and the rest are used to provide hover information. In the example above, your `ndarrays` would be for `name`, `tonnage`, `length` and `heading` (note the fact that dtypes can vary). 

Then you would need a hover tool with the ability to present this information: the default would hopefully be something like the mockup but you would want the ability to customize the format.

### Feature description

This feature request builds on top of the one made in https://github.com/bokeh/bokeh/issues/13200: in fact, that request (preserving the categorical information) is a prerequisite for what is suggested here.

In addition to preserving the category labels, you now also have the various dimension labels (e.g. you could pass in your ndarrays as a dictionary where the string keys are the names) e.g. `{'name:<string array>, 'tonnage': <float array>, ...}`

Finally, I would hope for a nice default hover tool behavior along the lines of the mockup (with setting to truncate when the list gets too long). However, the necessary extensions to the hover tool format specification should also be supported to allow custom hover tools that can present the categories, dimension labels and corresponding data in whatever format they desire.

Together with color mixing, I think such a system would be a very powerful way to explore categorical data in image stacks!

### Potential alternatives

For images, you could make a hover tool that flattens everything out into a list which becomes quite verbose and hard to read:

```
Cargo ship name: NORD VALIANT
Cargo ship tonnage: ...
Cargo ship length: ...
Cargo ship heading: ...
Passenger ship name: ORIENTAL STAR
Passenger ship tonnage: ...
Passenger ship length: ...
Passenger ship heading: ...
```

However, the goal would be to make image stacks work nicely.

*Edit*: Another potential alternative would be a hover tool that can communicate where you are hovering to the server and that can then display whatever information the server then computes in response (i.e. a dynamic version of the hover tool).

**Comment by jbednar** [comment]
That all sounds great. One addition to the "Potential alternatives" is to have only two ndarrays, one with the data being plotted, and one with the index of the data row corresponding to each pixel.  The index would then be used dynamically by a custom hover tool to query the original dataset for all of the other information being shown above, such that the data storage requirement only goes up by one ndarray regardless of how many columns in the dataset are exposed, instead of an ndarray for every column needed in the hover. (Additional ndarrays could give access to multiple index values (e.g. top `n` largest and top `n` smallest boats in this pixel), but still without storing the actual hover data per pixel.) 

The advantage of this approach is to save memory and the cost of assembling those arrays, under the assumption that most pixels (and probably most plots rendered) will never be hovered over. The main disadvantage is that hovering will be significantly slower each time, since it requires a round trip back to the server, while the proposal in this issue is all handled at the client side.  Another disadvantage is that a static HTML export will not show any of the hover information, without the server there to respond to the query.  

I can easily imagine situations where I'd prefer one behavior or the other (client-side or server-side hover data), so I'd argue for supporting both. HoloViews might be sufficient for the server-side support, but I think Bokeh would need to support the client-side approach laid out here.

**Comment by jlstevens** [comment]
> The main disadvantage is that hovering will be significantly slower each time, since it requires a round trip back to the server, while the proposal in this issue is all handled at the client side.

This is a good point to make. In HoloViews we have achieved this with having hover applied to a dynamically updating glyph containing data that changes as you hover around the image.

While this approach works, it is a bit of a hack. The ideal thing would be if there was some version of the Bokeh hover tool that could be directly updated from the server side without needing a proxy glyph. This isn't a feature I have ever proposed to Bokeh as I thought it might be considered out of scope. 

However, if the Bokeh devs do think that it might be good to be able to offer hover information populated dynamically from the server (instead of pushing all the data to the browser at once) then I would be happy to open another feature request!

*Edit*: I suppose that is another potential alternative! Note that @jbednar's suggestion of having an index layer is not strictly required on the client side. If there was a hover tool that could tell you the `x`, `y` position of the cursor (either pixel indices or in data space or both), you could look up (or compute) the hover information from that on the server side and then push this information back to the client to display in the hover.

**Comment by mattpap** [comment]
> However, if the Bokeh devs do think that it might be good to be able to offer hover information populated dynamically from the server (instead of pushing all the data to the browser at once) then I would be happy to open another feature request!

This sounds like a request for a lazy columnar data source, where data for requested indices is loaded as needed. This coupled with persistent tooltips and a loading indicator should result in a decent UI experience. Just to note, we already have "server sent" data sources, but those either load all data or work in append mode, so not really suited for this.

**Comment by ianthomas23** [comment]
@jlstevens Thanks for the clear write up, and the mockup is excellent. We can certainly work towards that, and an appropriate syntax for specifying how the tooltips are displayed plus sensible defaults.

There is one thing I would change, which is what data is and is not owned by an `image_stack` glyph. You propose passing in multiple 3D arrays of different dtypes to an `image_stack` with only one being used for colormapping and the rest for tooltips only. This doesn't really fit in with how the other `image`-like glyphs work in Bokeh which is that all the image data is numerical and used for colormapping, and you can pass in multiple images that are rendered at different positions and at different offsets. Nearly all use cases I have seen only use a single image of course, but we need to be consistent from an end-users point of view.

So instead you would put multiple 3D arrays of different dtypes into a single `ColumnDataSource`, and the `image_stack` uses the CDS and a named array from it for the colormapping. The remaining arrays are not known by the `image_stack` itself but are accessed from the tooltip code by the named keys which is consistent with how existing tooltips work. A benefit here is that you are not restricted to using 3D arrays as this approach supports other dimensionalities such as if you wanted to pass in a 4D array such as generated by a datashader `by(max_n)` reduction.

Actually, on reading that again it isn't really that different to what is proposed above if you consider the combination of CDS and `image_stack` as a single entity, which is often how client code works. I am maybe just being pedantic from a Bokeh internals point of view.

OK, let's be clear about the scope of the current work. I'll implement image stack labels (#13200) and then along with @mattpap will look into possible syntax for the new tooltips to achieve results like in the mockup above. We'll stick to 3D arrays (ignoring `max_n` etc) and client-side only hover data initially. It will probably be possible to include 4D arrays as it shouldn't be much extra work, but if that is holding us back I'd rather deliver the 3D array hover as soon as possible. Server-side hover data we'll put off as future research after all the client-side hover is working as desired.

**Comment by jbednar** [comment]
Sounds great, thanks!

**Comment by jbednar** [comment]
BTW, Jean-Luc's mockup is great but it's missing information about the counts of the categories, which I think is the most basic information, not needing any special server-side lookups. I.e., the ImageStack already has counts per category for every pixel, and at least by default we'd want to show that that pixe had 12 cargo ships and 2 passenger ships; the specific ships in the tooltip above are only _examples_, and we also want information about the distribution from which these examples were drawn. @jlstevens , not sure how best to fit that into the mockup...

**Comment by ianthomas23** [comment]
@jbednar That is already covered, you get that for free by us producing something like the mock up shows.

**Comment by jlstevens** [comment]
Right, in the final vision of how things work, `count` would typically be shown in the hover too and is the dimension used for color mixing. However, from the bokeh perspective, it is just another named dimension like any other: the important thing is to have the ability to show hover information for all available layers (whether or not they are involved in color mixing).

In other words, I should probably have shown `count` to represent the final goal we have but I don't think that doing so would have any affect on the feature request at the bokeh level.
